### PR TITLE
Add keyword argument support to activation checkpointing

### LINKED ISSUE #7038: [REQUEST] activation checkpoint API should have parity with Pytorch, keywords arguments not supported
**Is your feature request related to a problem? Please describe.**

`deepspeed.checkpointing.checkpoint` does not support keyword arguments. However `torch.utils.checkpoint.checkpoint` does support it, this makes it impossible to apply `deepspeed.checkpointing.checkpoint` as a direct replacement.

```python
from functools import partial

import torch
import deepspeed
import torch.utils.checkpoint

torch_checkpoint = partial(torch.utils.checkpoint.checkpoint, use_reentrant=False)

def fn(a, extra=None):
    if extra is not None:
        return a + extra
    else:
        return a

a = torch.tensor([1,2])
b = torch.tensor([3,4])

print(fn(a, extra=b)) # tensor([4, 6])
print(torch_checkpoint(fn, a, extra=b))  # tensor([4, 6])
print(deepspeed.checkpointing.checkpoint(fn, a, extra=b)) # Error,  checkpoint() got an unexpected keyword argument 'extra'
```

**Describe the solution you'd like**

Please add support for keyword arguments in `deepspeed.checkpointing.checkpoint`.

**Describe alternatives you've considered**

I could ensure that calls do not use keyword arguments. However, that is not feasible when the module is defined in a third-party library.

### PR DESCRIPTION
**Pull request by zupengwang**
## Summary

DeepSpeed's activation checkpointing wrapper currently accepts positional
arguments only, while `torch.utils.checkpoint.checkpoint` also supports keyword
arguments. This change allows callers to pass keyword arguments through
`deepspeed.checkpointing.checkpoint`.

Keyword names and non-Tensor values are retained for reconstruction during the
forward and recompute passes. Tensor keyword values are flattened into the
inputs passed to `CheckpointFunction`, so autograd tracks them and returns their
gradients correctly.

The activation checkpointing documentation now describes keyword argument
support, and the regression test covers both Tensor and non-Tensor keyword
arguments as well as gradient propagation.

## Validation

- Activation checkpointing unit tests: 27 passed
- Pre-commit checks for all changed files: passed
- 1-GPU CUDA correctness smoke: direct and checkpointed execution matched, with
  zero maximum gradient error
- 2-GPU distributed CUDA correctness smoke: each rank matched direct execution,
  with zero maximum gradient error
- DCO sign-off is included in the commit

Fixes #7038

### CODE REVIEW

**Comment by sfc-gh-truwase** [comment]
@zupengwang thanks for this PR. Looking forward to when it is ready for review.

File: deepspeed/runtime/activation_checkpointing/checkpointing.py [hunk 1]
```diff
@@ -945,9 +945,25 @@ def after_backward_hook(_nonuse_grads):
 
 
 @compiler.disable  # WA from Pytorch repo for compile + zero 3 accuracy issue
-def checkpoint(function, *args):
+def checkpoint(function, *args, **kwargs):
```
[the remark this answers is not in the document]
Comment (zupengwang) [reply deepspeed/runtime/activation_checkpointing/checkpointing.py:948]: Thanks for flagging this. The reviewed commit 47b2a888a6 includes the Signed-off-by: Wang Zupeng <zupenwang@gmail.com> trailer in its raw commit message, and its DCO check passed. The current rebased head 8d56205338 retains the same trailer and also passes DCO, so no additional commit rewrite is needed.

**Comment by zupengwang** [comment]
@sfc-gh-truwase Thanks! The PR is ready for review now. DCO and test collection have passed, and the remaining CI job is in progress.

**Review by sfc-gh-truwase — APPROVED** [review]
